lighthouse-core/test/mocha-setup.js

```diff
@@ -0,0 +1,148 @@
+
+/** @type {Map<string, SnapshotState['prototype']>} */
+const snapshotStatesByTestFile = new Map();
+
+/**
+ * @param {string} testFile
+ */
+function getSnapshotState(testFile) {
+  // For every test file, persist the same snapshot state object so there is
+  // not a read/write per snapshot access/change, but one per file.
+  let snapshotState = snapshotStatesByTestFile.get(testFile);
+  if (snapshotState) return snapshotState;
+
+  const snapshotDir = path.join(path.dirname(testFile), '__snapshots__');
+  const snapshotFile = path.join(snapshotDir, path.basename(testFile) + '.snap');
+  snapshotState = new SnapshotState(snapshotFile, {
+    updateSnapshot: process.env.SNAPSHOT_UPDATE ? 'all' : 'new',
+    prettierPath: '',
+    snapshotFormat: {},
+  });
+  snapshotStatesByTestFile.set(testFile, snapshotState);
+  return snapshotState;
+}
+
+/**
+ * @param {Mocha.Test} test
+ * @return {string}
+ */
+function makeTestTitle(test) {
+  /** @type {Mocha.Test | Mocha.Suite} */
+  let next = test;
+  const title = [];
+
+  for (;;) {
+    if (!next.parent) {
+      break;
+    }
+
+    title.push(next.title);
+    next = next.parent;
+  }
+
+  return title.reverse().join(' ');
+}
+
+expect.extend({
+  /**
+   * @param {any} actual
+   */
+  toMatchSnapshot(actual) {
+    const test = mochaCurrentTest;
+    if (!test.file) throw new Error('unexpected value');
+
+    const title = makeTestTitle(test);
+    const snapshotState = getSnapshotState(test.file);
+    /** @type {import('jest-snapshot/build/types').Context} */
+    // @ts-expect-error - this is enough for snapshots to work.
+    const context = {snapshotState, currentTestName: title};
+    const matcher = toMatchSnapshot.bind(context);
+
+    return matcher(actual);
+  },
+  /**
+   * @param {any} actual
+   * @param {any} expected
+   */
+  toMatchInlineSnapshot(actual, expected) {
+    const test = mochaCurrentTest;
+    if (!test.file) throw new Error('unexpected value');
+
+    const title = makeTestTitle(test);
+    const snapshotState = getSnapshotState(test.file);
+    /** @type {import('jest-snapshot/build/types').Context} */
+    // @ts-expect-error - this is enough for snapshots to work.
+    const context = {snapshotState, currentTestName: title};
+    const matcher = toMatchInlineSnapshot.bind(context);
+
+    return matcher(actual, expected);
+  },
+});
+
+// @ts-expect-error
+global.expect = expect;
+
+/**
+ * @param {Mocha.HookFunction} mochaFn
+ * @return {jest.Lifecycle}
+ */
+const makeFn = (mochaFn) => (fn, timeout) => {
+  mochaFn(function() {
+    // eslint-disable-next-line no-invalid-this
+    if (timeout !== undefined) this.timeout(timeout);
+
+    /** @type {jest.DoneCallback} */
+    const cb = () => {};
+    cb.fail = (error) => {
+      throw new Error(typeof error === 'string' ? error : error?.message);
+    };
+    return fn(cb);
+  });
+};
+
+const {before, after} = require('mocha');
+global.beforeAll = makeFn(before);
+global.afterAll = makeFn(after);
```

Comment:
Mocha does `global.before`, jest does `global.beforeAll`. To make transition slightly easier, I thought having the same interface for some parts would be helpful.